tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+#define STICKY_DFLT_KEY		"Tempesta FW Default Key"
+
+#define tfw_string(v)		{ (v), sizeof(v) - 1 }
+
+TfwCfgSticky tfw_cfg_sticky;
+static struct crypto_shash *tfw_sticky_shash;
+
+/*
+ * Build a Tempesta HTTP message from pieces of data.
+ *
+ * The functions tfw_http_msg_setup() and tfw_http_msg_add_data()
+ * are designed to work together. The objective is to avoid error
+ * processing when putting stream data in the SKBs piece by piece.
+ *
+ * Errors may be returned by memory allocation functions,
+ * so that part is done in tfw_http_msg_setup(). Given the total
+ * HTTP message length, it allocates an appropriate number of SKBs
+ * and page fragments to hold the payload, and sets them up in
+ * a Tempesta message.
+ *
+ * The SKBs are created complely headerless. The linear part of
+ * SKBs is set apart for headers, and stream data is placed in
+ * paged fragments. Lower layers will take care of prepending
+ * all necessary headers.
+ *
+ * tfw_http_msg_add_data() adds a piece of data to the message,
+ * forming a data stream piece by piece. All memory for the data
+ * has been allocated and set up by tfw_http_msg_setup(), so any
+ * errors that we may get are considered critical.
+ *
+ * State is kept between calls to these functions to facilitate
+ * quick access to current SKB and page fragment. State is passed
+ * and updated on each call to these functions.
+ */
+typedef struct tfw_msg_add_state {
+	struct sk_buff *skb;
+	unsigned int	fragnum;
+} tfw_mastate_t;
+
+static void
+tfw_http_msg_add_data(tfw_mastate_t *state, TfwMsg *msg, char *data, size_t len)
+{
+	skb_frag_t *frag;
+	struct sk_buff *skb = state->skb;
+	unsigned int i_frag = state->fragnum;
+	size_t copy_size, page_offset, data_offset = 0;
+
+	BUG_ON(skb == NULL);
+	BUG_ON(i_frag >= MAX_SKB_FRAGS);
+
+	while (len) {
+		if (i_frag >= MAX_SKB_FRAGS) {
+			skb = ss_skb_next(&msg->skb_list, skb);
+			state->skb = skb;
+			state->fragnum = 0;
+			i_frag = 0;
+			BUG_ON(skb == NULL);
+		}
+		for (; len && (i_frag < MAX_SKB_FRAGS); i_frag++) {
+			frag = &skb_shinfo(skb)->frags[i_frag];
+			page_offset = skb_frag_size(frag);
+			copy_size = min(len, PAGE_SIZE - page_offset);
+			memcpy(page_address(frag->page.p) + page_offset,
+			       data + data_offset, copy_size);
+			skb_frag_size_add(frag, copy_size);
+			data_offset += copy_size;
+			len -= copy_size;
+		}
+		/*
+		 * The above for() loop runs at least once,
+		 * which means that i_frags is always incremented.
+		 */
+		state->fragnum = i_frag - 1;
+	}
+	/* In the end, data_offset equals the initial len value */
+	skb->len += data_offset;
+	skb->data_len += data_offset;
+}
+
+static int
+tfw_http_msg_setup(tfw_mastate_t *state, TfwMsg *msg, size_t len)
+{
+	struct page *page;
+	struct sk_buff *skb;
+	int i_frag, i_skb, nr_skb_frags;
+	int nr_frags = DIV_ROUND_UP(len, PAGE_SIZE);
+	int nr_skbs = DIV_ROUND_UP(nr_frags, MAX_SKB_FRAGS);
+
+	/*
+	 * TODO: Make sure to create SKBs with payload size <= MSS
```

Comment:
Perhaps. After some digging in the Linux network stack code, I could see that whenever SKBs are filled  in the TCP stack, they are carefully made not to exceed MSS, including all headers. The same issue may arise when adding extra headers to an HTTP request, as that operation may cause the SKB payload to exceed MSS.

I guess this issue needs extra research, and it needs a clear written analysis of whats and whys.